File: backend/templates/workflow-engine/ai/classify.ts
/** AI classify handler — classifies input into labels. */

import { AINodeHandler } from './base';

export class ClassifyHandler extends AINodeHandler {
  async execute(config: Record<string, unknown>): Promise<Record<string, unknown>> {
    const aiInput = this.resolver.resolveString((config.aiInput as string) ?? '');
    const aiLabels = (config.aiLabels as string[]) ?? [];
    const aiPrompt = this.resolver.resolveString((config.aiPrompt as string) ?? '');
    const threshold = parseFloat(String(config.aiThreshold ?? 0.7));

    const labelsStr = aiLabels.length > 0 ? aiLabels.join(', ') : 'positive, negative, neutral';

    const systemPrompt =
      'You are a classification assistant. Classify the input into exactly one of the ' +
      `given labels. Labels: [${labelsStr}]. ` +
      'Respond with JSON: {"label": "<label>", "confidence": <0.0-1.0>}';
    const userPrompt = aiPrompt ? `${aiPrompt}\n\nInput to classify:\n${aiInput}` : aiInput;

    const response = await this.callLLM(systemPrompt, userPrompt);
    const parsed = this.parseJsonResponse(response);

    const label = (parsed.label as string) ?? (aiLabels[0] ?? 'unknown');
    const confidence = parseFloat(String(parsed.confidence ?? 0.8));

    return {
      label,
      confidence,
      meets_threshold: confidence >= threshold,
      output: label,
    };
  }

  protected override mockResponse(): string {
    return JSON.stringify({ label: 'positive', confidence: 0.85 });
  }
}

File: backend/runtime/ai/classify.py
"""AI classify handler — classifies input into labels."""

import json
import logging

from runtime.ai.base import AINodeHandler

logger = logging.getLogger(__name__)


class ClassifyHandler(AINodeHandler):
    """Classifies input text into one of the provided labels."""

    async def execute(self, config: dict) -> dict:
        ai_input = self.resolver.resolve_string(config.get("aiInput", ""))
        ai_labels = config.get("aiLabels", [])
        ai_prompt = self.resolver.resolve_string(config.get("aiPrompt", ""))
        threshold = float(config.get("aiThreshold", 0.7))

        labels_str = ", ".join(ai_labels) if ai_labels else "positive, negative, neutral"

        system_prompt = (
            "You are a classification assistant. Classify the input into exactly one of the "
            f"given labels. Labels: [{labels_str}]. "
            "Respond with JSON: {\"label\": \"<label>\", \"confidence\": <0.0-1.0>}"
        )
        user_prompt = f"{ai_prompt}\n\nInput to classify:\n{ai_input}" if ai_prompt else ai_input

        response = await self._call_llm(system_prompt, user_prompt)
        parsed = self._parse_json_response(response)

        label = parsed.get("label", ai_labels[0] if ai_labels else "unknown")
        confidence = float(parsed.get("confidence", 0.8))

        return {
            "label": label,
            "confidence": confidence,
            "meets_threshold": confidence >= threshold,
            "output": label,
        }

    def _mock_response(self, system_prompt: str, user_prompt: str) -> str:
        return json.dumps({"label": "positive", "confidence": 0.85})
